content/5_Plat/Range_Sweep.mdx

```diff
@@ -91,20 +93,58 @@ int main() {
 
 ## Solution - Springboards
 
-<IncompleteSection />
+**Naive DP:** $\mathcal{O}(P^2)$
 
-The problem boils down to having a data structure that supports the following
-operations:
+The first step is to create a DP to solve the first subtask. The states are the springboards and the transitions are between springboards. First, sort the springboards by the pair $(x_1, y_1)$ in increasing order. It is possible to show that for all $i$, $j$, where $i < j$, Bessie cannot use springboard $j$ then $i$ later.
 
-1. Add a pair $(a,b)$.
-2. For any $x$, query the minimum value of $b$ over all pairs satisfying
-   $a\le x$.
+For each springboard $i$, let $\texttt{ans}[i]$ denote the minimum distance needed to walk to the start point of springboard $i$.
+
+Let $\texttt{dist}(a, b)$ be the walking distance from the end of springboard $a$ and the start of springboard $b$.
+
+$$$
+\texttt{dist}(a, b) = x_1[b] + y_1[b] - x_2[a] - y_2[a]
+$$$
+
+Then, the transitions are:
+
+$$$
+\texttt{ans}[i] = \min\limits_{j < i, x_2[j] \le x_1[i], y_2[j] \le y_1[i]}(ans[j] + \texttt{dist}(j, i))
+$$$
+
+**Full Solution:** $\mathcal{O}(P \log P)$
+
+Optimizing the DP involves the use of a min point update range query segment tree. Let's first expand $dist(i, j)$ in the transition formula.
+
+$$$
+\texttt{ans}[i] = \min\limits_{j < i, x_2[j] \le x_1[i], y_2[j] \le y_1[i]}(\texttt{ans}[j] + x_1[i] + y_1[i] - x_2[j] - y_2[j])
+$$$
+$$$
+\texttt{ans}[i] = x_1[i] + y_1[i] + \min\limits_{j < i, x_2[j] \le x_1[i], y_2[j] \le y_1[i]}(\texttt{ans}[j] - x_2[j] - y_2[j])
+$$$
+
+We notice that everything inside the $\min$ statement only depends on $j$. The segment tree stores $\texttt{ans}[j] - x_2[j] - y_2[j]$ at index $y_2[j]$. We can seperate the start and end of springboards to create two seperate events for each springboard, still sorting by $(x, y)$. When the event is the start of a springboard, update $ans[i]$ through a segment tree query. When the event is the end of a springboard, update the segment tree.
 
-The [other module](/adv/springboards) describes a simpler way of doing this,
-though it's probably more straightforward to do coordinate compression and use a
-min segtree. The latter approach is shown below.
+By processing in order, the first two conditions in the $\min$ statement are always satisfied. The third is where the segment tree comes into play, where querying the range $[0, y_1[i]]$ is sufficent to satisfy all constraints.
+
+Due to the large $N$, coordinate compression is required in the code.
 
 ```cpp
+#include <bits/stdc++.h>
+using namespace std;
+#define f first
+#define s second
+#define pb push_back
+#define rep(i, a, b) for (int i = a; i < (b); ++i)
+#define all(x) begin(x), end(x)
+#define sz(x) (int)(x).size()
+typedef long long ll;
+typedef pair<int, int> pii;
+typedef vector<int> vi;
```

Comment:
can you not use these macros

Reply:
Ok I will change. Note that I am just using the original code on the website before.